CONTRIBUTING.md

```diff
@@ -0,0 +1,136 @@
+what it improves and how. Make sure the body lines fit in 72 characters,
+doesn't contain HTML, Markdown or any other markup language. Just plain text.
+
+Try putting yourself in a reviewer's shoes and see whether they can, without
+prior knowledge of the issue, understand the _why_, the _what_ and the _how_.
+Add any relevant benchmarks, performance and other information such as forums
+and mailing list discussions that helps in understanding and evaluating whether
+the change is the optimal way among alternatives.
+
+If the change resolves an existing issue, bugfix or a feature implementation,
+append to the body `Fixes #123` or `Closes #123` where `#123` is the issue
+number. If it's a partial resolution, use `Updates #123` or `Refs #123` instead.
+
+Here's an example of a good commit message:
+
+    docs: add contribution guidelines
+
+    As the contributors list keeps growing, we really need to get on the
+    same page in how to propose and make changes to the codebase.
+    The guidelines in this commit align with what people with merge
+    permissions to this repo consider common practices.
+
+    There are other git repos, such as bitbox02-firmware, api-go and api-js.
+    We'll need the same guidelines in those, too. An alternative was to
+    publish a single copy on a website but having independent
+    self-sufficient repos was a prevailing benefit. So, will copy these
+    guidelines over to the other repos once this is committed.
+
+    Closes #96
+
+See `git log` for more examples.
+
+For details on working with git, please refer to the
+[git manual](https://git-scm.com/doc).
+
+## Pull requests
+
+A pull request contains one or more related git commits. Please, do not bundle
+independent and unrelated commits into a single pull request.
+
+Just like a git commit message, a pull request consists of a subject and a body.
+If a pull request contains only one git commit, set its title and description to
+the commit's subject and the body. Otherwise, make an overall summary of what
+all the commits accomplish together, in a way similar to a commit message.
+
+Before creating a pull request, please make sure all tests and lints pass
+locally on our machine. In a case where a pull request isn't ready for
+a regular code review and you're just looking for some early feedback,
+it's ok to let some tests fail but please mention it explicitly in the request
+description. See readme file in the root of the repository for setting up
+development environment and building/testing instructions.
+
+At this stage one should expect comments and code reviews from other contributors.
+To address review comments, amend an existing commit with:
+
+    git commit --amend
+
+and overwrite remote branch with:
+
+    git push --force
+
+> **NOTE:** If you're collaborating on a pull request with others, it is not
+> recommended to overwrite the remote branch and simple add more commits on top.
```

Comment:
this reas as if it is not recommended to add more commits, but you mean the opposite i think.

Reply:
actually, i think it's too much detail here. i'll probably remove this note.

Reply:
removed the note.